### ISSUE TITLE
Unnecessary type repetition in trait bound

### ISSUE THREAD

**Issue by xd009642** [rust-lang/rust-clippy#3764]
So I raised this on https://github.com/rust-lang/rust/issues/58471 and it was suggested this is better as a clippy lint so pasting the issue text here for quick reference:

> My where clauses in traits recently had some unnecessary noise that the compiler could warn against:
> 
> What I was doing:
> ```Rust
> impl<T> SomeTrait<T> where T: Clone, T: Debug
> ```
> 
> What I should have been doing:
> ```Rust
> impl<T> SomeTrait<T> where T: Clone + Debug
> ```
> 
> I'd be happy to implement this myself if there's a consensus on adding this as a warning, though I may > need some mentoring :+1: 

The offer to help with a PR myself still stands, I had a quick check of the lints and didn't spot it.

**Comment by oli-obk** [comment]
There are theoretically infinte ways to write where bounds. So as a first step just doing unification of `T: Foo, T: Bar` to `T: Foo + Bar` seems like a good idea.

This lint should not trigger inside macros (as it doesn't matter there and some macros might get hard to write), so you should use the `in_macro` function to bail out if we're inside a macro.

You can implement the [`check_generics`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/lint/trait.LateLintPass.html#method.check_generics) method and access the `where_clauses` field on the generics argument. If you want any help with the implementation, don't hesitate to ask here.

**Comment by xd009642** [comment]
I've just done a draft pull request. It's roughly there and builds but doesn't yet work.
